arangod/Aql/ModificationExecutor2.cpp

```diff
@@ -0,0 +1,286 @@
+  // queries
+  if (_modifier.size() == 0 || _infos._options.silent) {
+    _modifier.setupIterator();
+    while (!_modifier.isFinishedIterator()) {
+      InputAqlItemRow row{CreateInvalidInputRowHint{}};
+      std::tie(std::ignore, row, std::ignore) = _modifier.getOutput();
+      output.copyRow(row);
+      _modifier.advanceIterator();
+    }
+  } else {
+    ModOperationType modOp;
+    InputAqlItemRow row{CreateInvalidInputRowHint{}};
+    VPackSlice elm;
+
+    // TODO: Make this a proper iterator
+    _modifier.setupIterator();
+    while (!_modifier.isFinishedIterator()) {
+      std::tie(modOp, row, elm) = _modifier.getOutput();
+
+      bool error = VelocyPackHelper::getBooleanValue(elm, StaticStrings::Error, false);
+      if (!error) {
+        switch (modOp) {
+          case ModOperationType::APPLY_RETURN: {
+            if (_infos._options.returnNew) {
+              AqlValue value(elm.get(StaticStrings::New));
+              AqlValueGuard guard(value, true);
+              output.moveValueInto(_infos._outputNewRegisterId, row, guard);
+            }
+            if (_infos._options.returnOld) {
+              auto slice = elm.get(StaticStrings::Old);
+              if (slice.isNone()) {
+                slice = VPackSlice::nullSlice();
+              }
+              AqlValue value(slice);
+              AqlValueGuard guard(value, true);
+              output.moveValueInto(_infos._outputOldRegisterId, row, guard);
+            }
+            if (_infos._doCount) {
+              stats.incrWritesExecuted();
+            }
+            break;
+          }
+          case ModOperationType::IGNORE_RETURN: {
+            output.copyRow(row);
+            if (_infos._doCount) {
+              stats.incrWritesIgnored();
+            }
+            break;
+          }
+          case ModOperationType::IGNORE_SKIP: {
+            output.copyRow(row);
+            if (_infos._doCount) {
+              stats.incrWritesIgnored();
+            }
+            break;
+          }
+          case ModOperationType::APPLY_UPDATE:
+          case ModOperationType::APPLY_INSERT: {
+            // These values should not appear here anymore
+            // As we handle them in the UPSERT modifier and translate them
+            // into APPLY_RETURN
+            TRI_ASSERT(false);
+          }
+          default: {
+            TRI_ASSERT(false);
+            break;
+          }
+        }
+      }
+    }
+    output.advanceRow();
+    _modifier.advanceIterator();
+  }
+}
+
+template <typename FetcherType, typename ModifierType>
+std::pair<ExecutionState, typename ModificationExecutor2<FetcherType, ModifierType>::Stats>
+ModificationExecutor2<FetcherType, ModifierType>::produceRows(OutputAqlItemRow& output) {
+  TRI_ASSERT(_infos._trx);
+
+  // TODO: isDBServer case
+  ExecutionState state;
+  ModificationExecutor2::Stats stats;
+
+  _modifier.reset();
+
+  size_t maxOutputs = std::min(output.numRowsLeft(), ExecutionBlock::DefaultBatchSize());
```

Comment:
You can rely on numRowsLeft() here.
The outer code makes sure that output is not larger then DefaultBatchSize()